File: day-06/custom_customs.py
input_file = 'day-06/input.txt'

def get_groups(input):
  return [group.split('\n') for group in input.split('\n\n')]

def part1(input):
  return sum([len(set.union(*[set(sub) for sub in group])) for group in get_groups(input)])

def part2(input): 
  return sum([len(set.intersection(*[set(sub) for sub in group])) for group in get_groups(input)])

if __name__ == "__main__":
  with open(input_file) as f:
    data = f.read()
  print("Part 1: ", part1(data))
  print("Part 2: ", part2(data))

File: day-06/custom_customs.go
package main

import (
	"fmt"
	"io/ioutil"
	"strings"
)

const input_file string = "./day-06/input.txt"

func ParseGroups(input string) [][]string {
	groups := [][]string{}
	groups_string := strings.Split(strings.TrimSpace(input), "\n\n")
	for _, group_string := range groups_string {
		group := strings.Split(group_string, "\n")
		groups = append(groups, group)
	}
	return groups
}

func Part1(input string) int {
	union_count := 0
	groups := ParseGroups(input)
	for _, group := range groups {
		character_count := make(map[rune]int)
		for _, person := range group {
			for _, char := range person {
				character_count[char]++
			}
		}
		union_count += len(character_count)
	}
	return union_count
}

func Part2(input string) int {
	intersection_count := 0
	groups := ParseGroups(input)
	for _, group := range groups {
		character_count := make(map[rune]int)
		for _, person := range group {
			for _, char := range person {
				character_count[char]++
			}
		}
		for _, count := range character_count {
			if count == len(group) {
				intersection_count++
			}
		}
	}
	return intersection_count
}

func main() {
	b, err := ioutil.ReadFile(input_file)
	if err != nil {
		fmt.Print(err)
	}
	data := string(b)

	fmt.Printf("Part 1: %d\n", Part1(data))
	fmt.Printf("Part 2: %d\n", Part2(data))
}
